devices/src/virtio/mmio.rs

```diff
@@ -553,42 +641,74 @@ mod tests {
+        assert_eq!(d.driver_status, 0);
 
-        for q in d.queues.iter_mut() {
-            q.size = 16;
-            q.ready = true;
+        set_driver_status(&mut d, DEVICE_ACKNOWLEDGE);
+        set_driver_status(&mut d, DEVICE_ACKNOWLEDGE | DEVICE_DRIVER);
+        assert_eq!(d.driver_status, DEVICE_ACKNOWLEDGE | DEVICE_DRIVER);
+
+        // invalid state transition should have no effect
+        set_driver_status(
+            &mut d,
+            DEVICE_ACKNOWLEDGE | DEVICE_DRIVER | DEVICE_DRIVER_OK,
+        );
+        assert_eq!(d.driver_status, DEVICE_ACKNOWLEDGE | DEVICE_DRIVER);
+
+        set_driver_status(
+            &mut d,
+            DEVICE_ACKNOWLEDGE | DEVICE_DRIVER | DEVICE_FEATURES_OK,
+        );
+        assert_eq!(
+            d.driver_status,
+            DEVICE_ACKNOWLEDGE | DEVICE_DRIVER | DEVICE_FEATURES_OK
+        );
+
+        let mut buf = vec![0; 4];
+        for q in 0..d.queues.len() {
+            d.queue_select = q as u32;
+            LittleEndian::write_u32(&mut buf[..], 16);
+            d.write(0x38, &buf[..]);
+            LittleEndian::write_u32(&mut buf[..], 1);
+            d.write(0x44, &buf[..]);
         }
         assert!(d.are_queues_valid());
         assert!(!d.device_activated);
 
         // Device should be ready for activation now.
 
-        let buf = vec![0; 4];
-
         // A couple of invalid writes; will trigger warnings; shouldn't activate the device.
         d.write(0xa8, &buf[..]);
         d.write(0x1000, &buf[..]);
         assert!(!d.device_activated);
 
-        // We pretend the device is already activated; activation related logic shouldn't be called
-        // in this situation, even for a valid write.
-        d.device_activated = true;
-        d.write(0x30, &buf[..]);
-        assert!(d.mem.is_some());
-        assert!(d.interrupt_evt.is_some());
-
-        // Ok, we stop pretending the device is activated.
-        d.device_activated = false;
-
-        // We issue this write again to actually activate the device.
-        d.write(0x30, &buf[..]);
+        set_driver_status(
+            &mut d,
+            DEVICE_ACKNOWLEDGE | DEVICE_DRIVER | DEVICE_FEATURES_OK | DEVICE_DRIVER_OK,
+        );
+        assert_eq!(
+            d.driver_status,
+            DEVICE_ACKNOWLEDGE | DEVICE_DRIVER | DEVICE_FEATURES_OK | DEVICE_DRIVER_OK
+        );
         assert!(d.device_activated);
 
         // A write which changes the size of a queue after activation; currently only triggers
-        // a warning path.
+        // a warning path and have no effect on queue state.
+        LittleEndian::write_u32(&mut buf[..], 0);
+        d.queue_select = 0;
         d.write(0x44, &buf[..]);
+        d.read(0x44, &mut buf[..]);
+        assert_eq!(LittleEndian::read_u32(&buf[..]), 1);
+
+        LittleEndian::write_u32(&mut buf[..], 0x8f);
+        d.write(0x70, &buf[..]);
+        assert_eq!(d.driver_status, 0x8f);
+        // TODO: assert!(d.device_activated);
```

Comment:
Why `assert!(d.device_activated)` again? Is there any path to de-activate an activated device?

Reply:
This is a reminder that there still two cases unhandled yet:
1) virtio device driver writes FALIED flags to driver_status field;
2) virtio device driver writes 0 to driver_status fields to reset device;

Still thinking about how to handle above two state transitions which may affect the device_activated flag. The block issue is that VirtioDevice::reset() is defined as optional and the net/block virtio backend driver doesn't support reset yet.

Reply:
Please create issues for outstanding TODOs and reference them in the codebase like:

```
// TODO (#xxx): issue-xxx-summary
```

Having said that, please be aware that we've yet found no customer need to support device reset.